rgw: add support for tagging and other conditionals in policy

theanalyst opened the pull request:
This is built atop of https://github.com/ceph/ceph/pull/16901 which is less intrusive (since it only requires params from requests and can probably be considered as L backport as well), adding support for existing object tags and other conditionals, Since this requires us to read the xattr of the object itself, and is not as cheap as evaluating other policy conditions which were already a part of request parameters, we look into the policy and see if the required conditional (currently ExistingObjectTag, but might apply for similar conditionals in ACLs) is actually a part of policy and required to be a part of the policy-env.

theanalyst commented:
very much WIP atm, just posting for some initial feedback

chardan commented on src/rgw/rgw_op.cc:
```diff
@@ -577,6 +581,43 @@ int rgw_build_object_policies(RGWRados *store, struct req_state *s,
 }
 
+void rgw_add_to_iam_environment(rgw::IAM::Environment& e, const std::string& key, const std::string& val){
+  e.emplace(std::piecewise_construct,
```
This will indeed (in C++11) probably skip a temporary, and LGTM-- but, consider also:
m.insert({k, v});
...or even:
m[k] = v;

...sliiiighly less efficient but probably a bit more readable (note a copy will be made in any case since these are const& parameters).

chardan commented on src/rgw/rgw_op.cc:
```diff
@@ -577,6 +581,43 @@ int rgw_build_object_policies(RGWRados *store, struct req_state *s,
+static int rgw_iam_add_tags_from_bl(struct req_state* s, bufferlist& bl){
+  int op_ret;
+  RGWObjTags tagset;
+  auto bliter = bl.begin();
```
It looks like this variable isn't really used? You might as well put everything here in the try block and get rid of it?

theanalyst replied:
makes sense, moving to the try block

theanalyst commented on src/rgw/rgw_iam_policy_keywords.gperf:
```diff
@@ -109,6 +109,7 @@ Null, TokenKind::cond_op, TokenID::Null, (uint64_t) Type::null, true, true
 #s3:authType, TokenKind::cond_key, TokenID::s3authType, (uint64_t) Type::string, true, false
 #s3:signatureAge, TokenKind::cond_key, TokenID::s3signatureAge, (uint64_t) Type::number, true, false
 #s3:x-amz-content-sha256, TokenKind::cond_key, TokenID::s3x_amz_content_sha256, (uint64_t) Type::string, true, false
+#s3:RequestObjectTag, TokenKind::cond_key, TokenID::s3RequestObjectTag, (uint64_t) Type::string, true, false
```
drop this, not evaluated anyway

theanalyst commented on src/rgw/rgw_iam_policy_keywords.h:
```diff
@@ -87,6 +87,7 @@ enum class TokenID {
   s3authType,
   s3signatureAge,
   s3x_amz_content_sha256,
+  s3RequestObjectTag,
```
drop this as well

theanalyst commented:
@adamemerson can you take a look and suggest if this is the direction to proceed (for requests that normally require reading object xattrs and such)

theanalyst commented:
barring versionid, and s3:RequestObjectTagKeys and we now have support for almost all of http://docs.aws.amazon.com/AmazonS3/latest/dev/amazon-s3-policy-keys.html#object-keys-in-amazon-s3-policies

adamemerson commented on src/rgw/rgw_iam_policy.cc:
```diff
@@ -1560,6 +1560,15 @@ Effect Policy::eval(const Environment& e,
+bool Policy::has_conditional(const string& conditional, bool partial) const {
+  for (const auto&s: statements){
+    for (const auto&c: s.conditions)
+      if (c.has_key(conditional, partial))
```
I'd use any_of here, but that's just me.

adamemerson replied:
Thinking for the future it might be worth having some condition keys be evaluated lazily, though that might make more sense in some future highly asynchronous RGW where eval can return a future.

chardan commented on src/rgw/rgw_op.cc:
```diff
@@ -590,7 +590,14 @@ void rgw_add_to_iam_environment(rgw::IAM::Environment& e, std::string&& key, std
 void rgw_add_to_iam_environment(rgw::IAM::Environment& e, const std::string& key, const std::string& val){
-  e[key] = val;
+  if (!val.empty())
+    e[key] = val;
```
Quick question: is it an error condition if val is empty? An expected result? If so, instead of silent failure, should we either throw if it's an error, or return some indication if it's expected?

adamemerson replied:
We should allow empty values. For s3:Prefix, for example, s3:Prefix being present and equal to the empty string is different from s3:Prefix being absent.

chardan commented on src/rgw/rgw_op.cc:
```diff
@@ -590,7 +590,14 @@ void rgw_add_to_iam_environment(rgw::IAM::Environment& e, std::string&& key, std
+
+
+void rgw_add_to_iam_environment(rgw::IAM::Environment& e, const char* key, const char* val){
+  if (key != nullptr && val != nullptr)
```
You're checking to see if the pointers are nullptr, but not *also* that the values they contain are empty as above.

theanalyst replied:
dropping the empty check, since empty values though rare are legal since policy allows for null conditionals which just check the presence of a key and not a value

chardan replied:
Did you also drop the empty check from the std::string& version?

theanalyst replied:
yes this is dropped now

chardan commented on src/rgw/rgw_op.cc:
```diff
@@ -4570,11 +4579,27 @@ int RGWCopyObj::verify_permission()
+                                     ARN(dest_obj));
+      if (e == Effect::Deny) {
+        return -EACCES;
+      } else if (e == Effect::Pass &&
```
If you are returning in the if, the else block is just complicating things.

theanalyst replied:
actually this logic confused me initially as well. e takes a third value Effect::Allow in which case we avoid all the if - else checks

theanalyst commented:
@adamemerson  can you take a look again, basically added support for copy obj. conditionals as well, and wrote up a table of our supported policies. Except for versioning, every other conditional is tested in https://github.com/ceph/s3-tests/pull/191 

I'll squash the commits after review.

theanalyst commented:
@adamemerson @cbodley can you take a look, I've rebased against master, fixed the doc build errors and added s3-test cases for nearly all the cases at https://github.com/ceph/s3-tests/pull/191

theanalyst commented:
jenkins test this please

cbodley commented on src/rgw/rgw_op.cc:
```diff
@@ -579,6 +583,73 @@ int rgw_build_object_policies(RGWRados *store, struct req_state *s,
+void rgw_add_to_iam_environment(rgw::IAM::Environment& e, std::string&& key, std::string&& val){
+  e.emplace(std::piecewise_construct,
+	    std::forward_as_tuple(key),
+	    std::forward_as_tuple(val));
```
`key` and `val` are lvalues in this statement, so they need to be cast with `std::move()` to enable move construction

and the `piecewise_construct` overload isn't necessary when both constructors take a single argument (see https://github.com/ceph/ceph/pull/18706)

theanalyst replied:
dropping this, there doesn't seem places (yet) where we can call std::move on both the key and value in the current patchset.

cbodley commented on src/rgw/rgw_op.cc:
```diff
@@ -579,6 +583,73 @@ int rgw_build_object_policies(RGWRados *store, struct req_state *s,
+  } catch (buffer::error& err) {
+    ldout(s->cct,0) << "ERROR: caught buffer::error, couldn't decode TagSet" << dendl;
+    op_ret= -EIO;
+    return op_ret;
```
`return -EIO;` directly and you don't need `op_ret`

cbodley commented on src/rgw/rgw_op.cc:
```diff
@@ -579,6 +583,73 @@ int rgw_build_object_policies(RGWRados *store, struct req_state *s,
+    {"HTTP_X_AMZ_GRANT_READ_ACP", "s3:x-amz-grant-acp"},
+    {"HTTP_X_AMZ_GRANT_WRITE_ACP", "s3:x-amz-grant-write-acp"},
+    {"HTTP_X_AMZ_GRANT_FULL_CONTROL", "s3:x-amz-grant-full-control"},
+  };
```
this is expensive to construct every time! make it static, the string literals can stay `const char*`, and you can iterate over the `initializer_list` itself instead of constructing anything out of it:
```c++
  using header_pair = std::pair<const char*, const char*>;
  static const std::initializer_list<header_pair> acl_header_conditionals {
    {"HTTP_X_AMZ_GRANT_READ", "s3:x-amz-grant-read"},
    ...
  };
```

adamemerson commented on doc/radosgw/bucketpolicy.rst:
```diff
@@ -120,10 +120,83 @@ Currently, the only condition keys we support are:
+|s3:DeleteOBjectVersionTagging|                                               |                   |
++-----------------------------+-----------------------------------------------+-------------------+
+
 
```
Thank you. This is good documentation and I like it.

adamemerson commented on src/rgw/rgw_op.cc:
```diff
@@ -4533,11 +4533,27 @@ int RGWCopyObj::verify_permission()
     return op_ret;
   }
-
+  boost::optional <Policy> dest_iam_policy = get_iam_policy_from_attr(s->cct, store, dest_attrs, dest_bucket.tenant);
```
Auto? Seems better than having gigantic type specifiers and it will still work if we ever move to C++17 and std::optional.

chardan replied:
That will be a glorious day!

adamemerson commented on src/rgw/rgw_op.cc:
```diff
@@ -590,7 +590,14 @@ void rgw_add_to_iam_environment(rgw::IAM::Environment& e, std::string&& key, std
+}
+
+
+void rgw_add_to_iam_environment(rgw::IAM::Environment& e, const char* key, const char* val){
```
I'd be tempted to merge these again and just use boost::string_view

chardan replied:
Now the project is C++17, so there is std::string_view...

adamemerson commented on src/rgw/rgw_iam_policy.cc:
```diff
@@ -1535,8 +1535,8 @@ Effect Policy::eval(const Environment& e,
   for (const auto&s: statements){
-    for (const auto&c: s.conditions)
-      if (c.has_key(conditional, partial))
+    if (std::any_of(s.conditions.begin(), s.conditions.end(),
```
"Yay!" 
       — S. Weil

theanalyst commented:
@adamemerson @cbodley addressed the comments so far, sorry for the delay

liewegas commented:
any progress here?

theanalyst commented:
rebased this against current master and also the s3-tests at https://github.com/ceph/s3-tests/pull/191; @cbodley @adamemerson

cbodley commented on src/rgw/rgw_iam_policy.h:
```diff
@@ -446,6 +455,8 @@ struct Policy {
 	      boost::optional<const rgw::auth::Identity&> ida,
 	      std::uint64_t action, const ARN& resource) const;
+
+  bool has_conditional(const string& conditional, bool partial=true) const;
```
`partial=true` makes sense for use with `s3:ExistingObjectTag/`, but i tend to think that exact matches should be the default, with partial matches being opt-in. just thinking about new callers in the future using `has_conditional("foo")` without realizing that there's a `foobar` condition key also

theanalyst replied:
ack makes sense

chardan replied:
Not "required", but something to consider: what if you provided your default behavior, then an overload that took a predicated (which you would always call with the same parameters)?
template <typename MatchPred>
bool has_conditional(const std::string_view key_, MatchPred p) { return p(key, key_); } 
...and implement the other in terms of it.

chardan commented on src/rgw/rgw_iam_policy.cc:
```diff
@@ -1554,6 +1554,15 @@ Effect Policy::eval(const Environment& e,
+bool Policy::has_conditional(const string& conditional, bool partial) const {
+  for (const auto&s: statements){
+    if (std::any_of(s.conditions.begin(), s.conditions.end(),
+		    [&](Condition c) { return c.has_key(conditional, partial);}))
```
+1

chardan commented on src/rgw/rgw_op.cc:
```diff
@@ -562,6 +566,67 @@ int rgw_build_object_policies(RGWRados *store, struct req_state *s,
 }
 
+void rgw_add_to_iam_environment(rgw::IAM::Environment& e, const std::string& key, const std::string& val){
+    e[key] = val;
```
Not that this creates OR updates, so I think it's misleading WRT the same of the member function.

chardan commented on src/rgw/rgw_iam_policy.h:
```diff
@@ -363,6 +363,13 @@ struct Condition {
     }
   };
 
+  struct ci_starts_with {
```
It's fine as-written, but you could also just make this an ordinary function, since it's not tracking any data.

chardan commented on src/rgw/rgw_iam_policy.h:
```diff
@@ -446,6 +458,24 @@ struct Policy {
 	      std::uint64_t action, const ARN& resource) const;
+
+  template <typename F>
+  bool has_conditional_p(const string& conditional, F p) const{
```
This name is fine for me, but you could also just make it another overload of has_conditional(). 
If you find yourself back in this file, please add a space after "const".

chardan commented on src/rgw/rgw_iam_policy.h:
```diff
@@ -446,6 +458,24 @@ struct Policy {
+    return false;
+  }
+
+  bool has_conditional(const string& c) const {
```
NIce!

theanalyst commented:
updated, there was a bug in order of arguments which we pass via `has_key` since boost::istarts_with expects the test string as teh second argument. Fixed that, s3 policy tests are green after this.

cbodley commented:
s3tests in https://github.com/ceph/s3-tests/pull/191 merged - we'll want to pull those into s3-tests/ceph-master once this merges

yuriw commented:
![](https://github.trello.services/images/mini-trello-icon.png) [wip-yuri-testing-2018-02-14-2132](https://trello.com/c/TRo1V4u8/262-wip-yuri-testing-2018-02-14-2132)
